### ISSUE TITLE
Double signup with different passwords returns token

### ISSUE THREAD

**Issue by Sourcetreehugger** [keratin/authn-server#59]
I found an issue that caused me confusion. First I thought that it was a front-end issue, but it is in the back end, and the front end will report success even though the passwords are different.

To reproduce:

Step 1
Sign up using credentials: 
Username: User1234
Password: 9oiajshdfioudewiuufjhi8127y4981

Expect: Token (201)
Result: Token (201)

Step 2: 
Sign up using credentials: 
Username: User1234
Password: sometotallydifferentPassword1897387214

Expect: TAKEN 
Result: Token (201)

**Comment by cainlevy** [comment]
Interesting. Would you describe how you're recreating this issue? For example, are you issuing requests through authn-js, curl, or Postman?

Also, what database engine are you using?

**Comment by cainlevy** [comment]
Think you might be able to connect directly to the database and poke around? Does it have multiple accounts with the same username?

**Comment by Inigo4562** [comment]
Yes, I was already checking the database and there are not new rows being created with the same name.

**Comment by Sourcetreehugger** [comment]
@cainlevy  , Inigo is the backend dev of this project. I think we can share the repo with you if you are interested.

**Comment by Inigo4562** [comment]
So I have just run a small test using mysql instead cockroachdb and the problem seems to be gone. Do you have an idea why cockroachdb causes this problem?

**Comment by Mohammed90** [comment]
@cainlevy , do you think we're hitting a bug with lib/pq or sqlx? The Create code mirrors the code for sqlite and mysql, except for the PostgreSQL specific changes. From what I remember, I didn't face this issue with PostgreSQL while testing, but I'll check again later today.

**Comment by Mohammed90** [comment]
I've just verified with postgres:10, and I consistently receive the following JSON when I try to signup using the same username but different passwords:
```json
{
   "errors":[
      {
         "field":"username",
         "message":"TAKEN"
      }
   ]
}
```

However, with CockroachDB, I get the same erroneous behavior reported by @Sourcetreehugger.

**Comment by cainlevy** [comment]
> I believe this means CockroachDB isn't returning an error to the driver when it should, right?

Yes, this is my best guess. Maybe it's returning a different error code than what's expected here: https://github.com/keratin/authn-server/blob/master/data/data.go#L71

**Comment by Mohammed90** [comment]
Nope, it doesn't return an error at all. Here's a code snippet only for CockroachDB where it only returns new IDs, no errors are printed, whatsoever:

```go
package main

import (
	"database/sql"
	"log"

	"github.com/lib/pq"
)

const (
	CREATE_TABLE_KEY = "create_table"
	INSERT_USER_KEY  = "insert_user"
)

func main() {

	sqlOps := map[string]string{
		CREATE_TABLE_KEY: `
		CREATE TABLE IF NOT EXISTS myTable (
			id SERIAL PRIMARY KEY,
			username TEXT UNIQUE DEFAULT NULL
		)`,
		INSERT_USER_KEY: `INSERT INTO myTable (username) VALUES ($1) RETURNING id`,
	}

	roachDB, roachDBOpenErr := sql.Open("postgres", "postgres://postgres@127.0.0.1:26257/postgres?sslmode=disable")
	if roachDBOpenErr != nil {
		log.Fatalf("roachDB: Couldn't open db: %s", roachDBOpenErr)
	}
	defer func() {
		err := roachDB.Close()
		if err != nil {
			log.Printf("roachDB close err is not nil: %s", err)
		}
	}()

	_, roachExecErr := roachDB.Exec(sqlOps[CREATE_TABLE_KEY])
	if roachExecErr != nil {
		log.Printf("roachExecErr is not nil: %s", roachExecErr)
		return
	}

	// insert first user
	{
		roachRows, roachQRowErr := roachDB.Query(sqlOps[INSERT_USER_KEY], "the_user")
		if roachQRowErr != nil {
			log.Printf("roachQRowErr is not nil: %s", roachQRowErr)
			return
		}
		if !roachRows.Next() {
			log.Printf("roachRows has no next on first insert")
			return
		}

		var firstRoachID int
		roachRows.Scan(&firstRoachID)

		log.Printf("First insert has ID: %d", firstRoachID)
		roachRows.Close()
	}

	// insert second user; expecting error
	{
		roachRows, roachQRowErr := roachDB.Query(sqlOps[INSERT_USER_KEY], "the_user")
		if roachQRowErr != nil {
			log.Printf("roachQRowErr is not nil: %s", roachQRowErr)
			if err, ok := roachQRowErr.(*pq.Error); ok {
				log.Printf("The error name is: %s", err.Code.Class().Name())
			}
			return
		}
		errVal := roachRows.Err()
		if errVal != nil {
			log.Printf("errVal is not nil: %s", errVal)
			if err, ok := errVal.(*pq.Error); ok {
				log.Printf("The error name is: %s", err.Code.Class().Name())
			}
			return
		}

		if !roachRows.Next() {
			log.Printf("roachRows has no next on second insert")
			return
		}

		var secondRoachID int
		roachRows.Scan(&secondRoachID)

		log.Printf("Second insert has ID: %d", secondRoachID)

		roachRows.Close()
	}

}
```

Complete Output:
```log
2018/07/30 19:49:06 First insert has ID: 369947267340369921
2018/07/30 19:49:06 Second insert has ID: 369947267353280513
```

The ID of the user in the database is the ID generated the first time. This means the second insert returns an ID with no errors, but no writes occured.

**Comment by cainlevy** [comment]
Nice work on that reproduction. I agree that this is likely a bug upstream in CockroachDB. Please reopen this bug if you learn otherwise.

**Comment by Mohammed90** [comment]
@Sourcetreehugger have you reported this to CockroachDB or should I report it myself? 
It seems like CockroachDB's SQL shell returns an error, but it doesn't communicate the same with the driver.

**Comment by Mohammed90** [comment]
I'll just leave this note as heads-up for whoever encounters this issue:

This particular issue is fixed in CockroachDB 2.1, which is in alpha release-stage at the time of this comment.

Reference:
https://github.com/cockroachdb/cockroach/issues/28150#issuecomment-409766892
